WalletWasabi.Fluent/ApplicationStateManager.cs

```diff
@@ -0,0 +1,263 @@
+			.Permit(Trigger.Minimised, State.Hidden)
+			.OnEntry(() =>
+			{
+				if (_lifetime.MainWindow is null)
+				{
+					CreateAndShowMainWindow();
+				}
+				else if (_lifetime.MainWindow.WindowState == WindowState.Minimized)
+				{
+					_lifetime.MainWindow.WindowState = WindowState.Normal;
+				}
+
+				if (ApplicationViewModel is { })
+				{
+					ApplicationViewModel.IsMainWindowShown = true;
+				}
+			});
+
+		_stateMachine.Configure(State.Hidden)
+			.SubstateOf(State.StandardMode)
+			.Permit(Trigger.Show, State.Shown)
+			.Permit(Trigger.Restored, State.Shown)
+			.Permit(Trigger.ShutdownPrevented, State.Shown)
+			.OnEntry(() =>
+			{
+				_lifetime.MainWindow.WindowState = WindowState.Minimized;
+
+				if (ApplicationViewModel is { })
+				{
+					ApplicationViewModel.IsMainWindowShown = false;
+				}
+			});
+
+		_lifetime.ShutdownRequested += LifetimeOnShutdownRequested;
+
+		Services.UiConfig.WhenAnyValue(x => x.HideOnClose)
+			.ObserveOn(RxApp.MainThreadScheduler)
+			.Subscribe(backgroundMode =>
+			{
+				if (backgroundMode)
+				{
+					_stateMachine.Fire(Trigger.BackgroundModeOn);
+				}
+				else
+				{
+					_stateMachine.Fire(Trigger.BackgroundModeOff);
+				}
+			});
+
+		_stateMachine.Start();
+
+		if (!startInBg)
+		{
+			_stateMachine.Fire(Trigger.Loaded);
+		}
+	}
+
+	private void MainWindowOnClosing(object? sender, CancelEventArgs e)
+	{
+		if (_stateMachine.IsInState(State.StandardMode))
+		{
+			if (ApplicationViewModel is { })
+			{
+				e.Cancel = !ApplicationViewModel.CanShutdown();
+			}
+
+			if (e.Cancel)
+			{
+				_stateMachine.Fire(Trigger.ShutdownPrevented);
+			}
+			else
+			{
+				_stateMachine.Fire(Trigger.ShutdownRequested);
+			}
+		}
+	}
+
+	private void LifetimeOnShutdownRequested(object? sender, ShutdownRequestedEventArgs e)
+	{
+		if (ApplicationViewModel is { })
+		{
+			// Shutdown prevention will only work if you directly run the executable.
+			e.Cancel = !ApplicationViewModel.CanShutdown();
+
+			Logger.LogDebug($"Cancellation of the shutdown set to: {e.Cancel}.");
+		}
+
+		if (e.Cancel)
+		{
+			_stateMachine.Fire(Trigger.ShutdownPrevented);
+		}
+		else
+		{
+			_stateMachine.Fire(Trigger.ShutdownRequested);
+		}
+	}
+
+	private void CreateAndShowMainWindow()
+	{
+		if (_lifetime.MainWindow is { })
+		{
+			return;
+		}
+
+		var result = new MainWindow
+		{
+			DataContext = MainViewModel.Instance
+		};
+
+		result.Closing += MainWindowOnClosing;
+
+		_subscriptions = result.WhenAnyValue(x => x.WindowState)
+			.Subscribe(x =>
+			{
+				if (x == WindowState.Minimized)
+				{
+					_stateMachine.Fire(Trigger.Minimised);
+				}
+				else
+				{
+					_stateMachine.Fire(Trigger.Restored);
+				}
+			});
+
+		Observable.FromEventPattern(result, nameof(result.Closed))
+			.Take(1)
+			.Subscribe(x =>
+			{
+				_subscriptions?.Dispose();
+				_subscriptions = null;
+				result.Closing -= MainWindowOnClosing;
+				_stateMachine.Fire(Trigger.MainWindowClosed);
+			});
```

Comment:
Wouldn't be simpler to introduce a `CompositeDisposable` and handle the cancellation of the subscriptions with that?

```c#
private CompositeDisposable? _disposable;
```

```c#
_disposable?.Dispose();
_disposable = new CompositeDisposable();

result.WhenAnyValue(x => x.WindowState)
	.Subscribe(x => _stateMachine.Fire(x == WindowState.Minimized ? Trigger.Minimised : Trigger.Restored))
	.DisposeWith(_disposable);

Observable.FromEventPattern<CancelEventArgs>(result, nameof(result.Closing))
	.Subscribe(x => MainWindowOnClosing(x.Sender, x.EventArgs))
	.DisposeWith(_disposable);

Observable.FromEventPattern(result, nameof(result.Closed))
	.Subscribe(_ => _stateMachine.Fire(Trigger.MainWindowClosed))
	.DisposeWith(_disposable);
```